### PR TITLE
Add `@tool_button` annotation for easily creating inspector buttons.

### LINKED ISSUE #2149: Add a way to export clickable actions to the inspector
*See https://github.com/godotengine/godot/issues/9380 for previous discussion.*

### Describe the project you are working on

The Godot editor :slightly_smiling_face: 

### Describe the problem or limitation you are having in your project

There's no officially recognized/documented way to export a clickable action to the inspector. Such actions can be useful to expose "one-off" things you may need to do when using a node. This can be useful to expose additional functionality without having to create an editor plugin.

There's a [hack](https://github.com/godotengine/godot-docs/pull/4106) you can use that works with boolean exported properties, but it's not particularly user-friendly.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a way to export proper clickable buttons in the Inspector, without exporting dummy properties.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add an annotation that exports a method as a clickable action in the inspector. Like for property names, the action's name is named automatically based on the method name. For instance:

```gdscript
@export_func
func bake_all():
    pass
```

This would result in a clickable "Bake All" button in the inspector.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes (see above), but it's not particularly clean or user-friendly. A checkbox isn't an one-state button :slightly_smiling_face: 

### Is there a reason why this should be core and not an add-on in the asset library?

This is core editor functionality.

### PR DESCRIPTION
**Pull request by fire**
Supersedes: https://github.com/godotengine/godot/pull/59289
Closes: https://github.com/godotengine/godot-proposals/issues/2149

tool_button demo

- [x] I see you replaced UndoRedo with EditorUndoRedoManager in response to my comment. Unfortunately it won't work, because all actions should go through EditorUndoRedoManager.
- [ ] Need to test api changes

### CODE REVIEW

**Comment by dalexeev** [comment]
`TOOL_BUTTON` as well as `GROUP` should be a "pseudo member" and must not shadow real members in `members_indices`, see #78254. In property list the names should be different too, I think (unlike `GROUP`).

To change the order of a tool button in the inspector, are you forced to move the function up? This does not follow the GDScript style guide, which recommends that variables be placed above functions.

**Comment by YuriSizov** [comment]
> To change the order of a tool button in the inspector, are you forced to move the function up? This does not follow the GDScript style guide, which recommends that variables be placed above functions.

The last time we discussed the syntax it was suggested that you can add the annotation wherever you want if you pass a callable "reference" to it, see https://github.com/godotengine/godot/pull/59289#issuecomment-1155140377.

**Comment by fire** [comment]
Can you explain how this affect this pr?

File: editor/plugins/tool_button_editor_plugin.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+class ToolButtonEditorPlugin : public EditorPlugin {
+	GDCLASS(ToolButtonEditorPlugin, EditorPlugin);
+
+public:
+	virtual String get_name() const override { return "ToolButtonEditorPlugin"; }
+
+	ToolButtonEditorPlugin();
+};
+
+#endif // TOOL_BUTTON_EDITOR_PLUGIN_H
```
Comment (AThousandShips) [inline editor/plugins/tool_button_editor_plugin.h:57]: Missing line termination

**Comment by fire** [comment]
Can someone help me test this? Also, does anyone know how to call this feature from c++?

File: modules/gdscript/gdscript_parser.cpp [hunk 2]
```diff
@@ -4174,6 +4178,44 @@ bool GDScriptParser::warning_annotations(const AnnotationNode *p_annotation, Nod
 #endif // DEBUG_ENABLED
 }
 
+bool GDScriptParser::tool_button_annotation(const AnnotationNode *p_annotation, Node *p_node) {
+#ifndef TOOLS_ENABLED // !TOOLS_ENABLED
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_parser.cpp:4182]: I'd flip the condition here, with both cases having the negative condition *first* is confusing

**Comment by ThomasUijlen** [comment]
Any news on this PR? This would be fantastic when creating plugins or editor tools.

**Comment by WideArchShark** [comment]
I too would love an update on this. Just downloaded 4.3-dev2 like a muppet thinking it would be in there!

File: modules/gdscript/doc_classes/@GDScript.xml [hunk 3]
```diff
@@ -655,6 +655,30 @@
...
+				    var random_color = Color(randf_range(0, 1.0), randf_range(0, 1.0), randf_range(0, 1.0));
+				    call_action(self, do_method_name, undo_method_name);
+
+				func set_color(new_color):
+				    modulate = new_color
+
+				func reset_color(old_color):
+				    modulate = old_color
+				[/codeblock]
+				[b]Note:[/b] Make sure to provide separate `do` and `undo` methods that perform and reverse the action respectively.
```
Comment (Calinou) [inline modules/gdscript/doc_classes/@GDScript.xml:679]:
```suggestion
				[b]Note:[/b] Make sure to provide separate [code]do[/code] and [code]undo[/code] methods that perform and reverse the action respectively.
```

File: modules/gdscript/gdscript_parser.cpp [hunk 4]
```diff
@@ -4174,6 +4178,44 @@ bool GDScriptParser::warning_annotations(const AnnotationNode *p_annotation, Nod
...
+		return false;
+	}
+	String hint_string = vformat("%s,%s", annotation->resolved_arguments[0], func->identifier->name);
+	if (annotation->resolved_arguments.size() > 1) {
+		// Icon.
+		hint_string += "," + annotation->resolved_arguments[1].operator String();
+	}
+	annotation->export_info.hint_string = hint_string;
+	return true;
+#endif // TOOLS_ENABLED
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_parser.cpp:4216]:
```suggestion
#ifdef TOOLS_ENABLED
	AnnotationNode *annotation = const_cast<AnnotationNode *>(p_annotation);
	FunctionNode *func = static_cast<FunctionNode *>(p_node);
	if (annotation->resolved_arguments.size() < 1) {
		push_error("Tool buttons must specify a name.", p_annotation);
		return false;
	}

	if (!this->is_tool()) {
		push_error("Tool buttons can only be used in tool scripts.", p_annotation);
		return false;
	}

	annotation->export_info.name = annotation->resolved_arguments[0];
	annotation->export_info.type = Variant::Type::CALLABLE;
	annotation->export_info.usage = PROPERTY_USAGE_EDITOR | PROPERTY_USAGE_INTERNAL;
	if (func->parameters.size() != 1) {
		push_error("Tool button methods must have an UndoRedo as their only argument.", func);
		return false;
	}
	if (func->parameters[0]->datatype_specifier != nullptr && func->parameters[0]->datatype_specifier->type_chain[0]->name != "UndoRedo") {
		push_error("Tool button methods must have an UndoRedo as their only argument.", func);
		return false;
	}
	String hint_string = vformat("%s,%s", annotation->resolved_arguments[0], func->identifier->name);
	if (annotation->resolved_arguments.size() > 1) {
		// Icon.
		hint_string += "," + annotation->resolved_arguments[1].operator String();
	}
	annotation->export_info.hint_string = hint_string;
	return true;
#else // !TOOLS_ENABLED
	// Only available in editor.
	return true;
#endif // TOOLS_ENABLED
```
As per the earlier suggestion

**Comment by caphindsight** [comment]
Is this still on track to make 4.3?

**Comment by AThousandShips** [comment]
No it's been abandoned by the author and needs to be salvaged, so it's waiting for someone to pick it up and continue the work

**Comment by caphindsight** [comment]
I might be insterested in picking it up.

Can you please summarize for me why the (arguably, overcomplicated) design with do/undo was chosen?
For my use cases, these actions I want `@tool_button` to perform are approximately never easily undo-able.
For example, I'd have a `build_mesh` action which would start by dropping the old mesh, and just building a new one instead, etc.

Wouldn't it be much simpler *and* more user-friendly to give up on undo?

```gdscript
@export var mesh_parameter: int = 42

@tool_button("Build mesh.")
func build_mesh():
    # Use mesh_parameter here, no function arguments allowed
```

This is exactly what people use the "boolean hack" for these days, only nicer and without the ugly workaround code.

**Comment by TokisanGames** [comment]
Undo/redo functionality would be used by a plugin or tool developer. `Click an inspector button to process the mesh. Oops, undo.` The tool developer has the option to connect their do and undo functions so the action gets logged in Godot's history and users can use them. It should also give the developer the option of providing a null undo callback so the functionality can be ignored by devs making hacks for their own use, IMO. But then these folks don't really need a tool button, as the boolean hack will suffice.

**Comment by Pshy0** [comment]
This feature could be added without undo/redo concerns by using a confirm dialogue.

**Comment by ultrasuperpingu** [comment]
> Undo/redo functionality would be used by a plugin or tool developer.

I'm not sure to understand why plugin or tool developper can't use the regular `UndoRedo` class to implements this if they need it ?

**Comment by TokisanGames** [comment]
You can. You're going to provide do and undo functions either way. This uses undo/redo in the background, so it's just more concise, syntactic sugar.

**Comment by ultrasuperpingu** [comment]
> You can. You're going to provide do and undo functions either way. This uses undo/redo in the background, so it's just more concise, syntactic sugar.

Ok, so, IMHO it's not a good idea to make "regular" user to have to provide a null undo callback as they probably would be the main users of this feature... On my addons, I'm personally not really bothered to have to use a specific UndoRedo API to handle this pretty advanced feature (and for now, I mainly used an EditorPlugin, EditorNode3DPlugin and EditorNode3DGizmoPlugin to perform undoable actions).

**Comment by KoBeWi** [comment]
I think the solution for undoredo problems is just exposing EditorUndoRedoManager via EditorInterface (which is accessible as a singleton). This way the users would be able to optionally provide undo for their action, without making button callbacks more complex. I'm already exposing it in #90130.

**Comment by Macksaur** [comment]
I salvaged the great work here in #96290 to fit my use-cases. If anyone would like to check it out and let me know if I've missed any obvious wins, that'd be swell.

**Comment by AThousandShips** [comment]
Superseded by:
* https://github.com/godotengine/godot/pull/96290

**Pull request closed without merging**